backend/common/src/main/java/ai/verta/modeldb/common/CommonUtils.java

```diff
@@ -172,6 +175,64 @@ public static <T extends GeneratedMessageV3> void observeError(
     responseObserver.onError(logError(e));
   }
 
+  public static <T> void observeError(CompletableFuture<T> completableFuture, Throwable e) {
+    completableFuture.completeExceptionally(observeErrorWithResponse(e, e.getMessage(), LOGGER));
+  }
+
+  public static ResponseStatusException observeErrorWithResponse(
+      Throwable e, String message, Logger logger) {
+    if (e instanceof ModelDBException
+        && e.getCause() instanceof ExecutionException
+        && e.getCause().getCause() != null) {
+      return observeErrorWithResponse(e.getCause().getCause(), message, logger);
+    }
+
+    if (e instanceof CompletionException) {
+      return observeErrorWithResponse(e.getCause(), message, logger);
+    }
+    final HttpStatus code;
+    if (e instanceof StatusRuntimeException) {
+      StatusRuntimeException ex = (StatusRuntimeException) e;
+      code = ModelDBException.httpStatusFromCode(ex.getStatus().getCode());
+      if (ex.getStatus().getCode() == io.grpc.Status.Code.NOT_FOUND) {
+        logger.debug(e.getMessage());
+      } else {
+        logger.error(e.getMessage());
+      }
+      message += " Details: " + e.getMessage();
+      printStackTrace(logger, e);
+    } else if (e instanceof IllegalArgumentException) {
+      code = HttpStatus.BAD_REQUEST;
+      String logMessage = "Common Service Exception occurred: {}";
+      message = e.getMessage();
+      logger.debug(logMessage, message);
+    } else if (e instanceof ModelDBException) {
+      ModelDBException uacServiceException = (ModelDBException) e;
+
+      code = uacServiceException.getHttpCode();
+      message += " Details: " + e.getMessage();
+      String logMessage = "Common Service Exception occurred: {}";
+      if (uacServiceException.getCodeValue() == Code.INTERNAL_VALUE) {
+        // If getting 500 then we will not expose about error to the user.
+        message = "Internal Server Error";
+        logger.error(logMessage, e.getMessage());
+        printStackTrace(logger, e);
+      } else {
+        if (uacServiceException.getCodeValue() == Code.RESOURCE_EXHAUSTED_VALUE) {
+          message = e.getMessage();
+        }
+        logger.debug(logMessage, e.getMessage());
+      }
+    } else {
+      code = HttpStatus.INTERNAL_SERVER_ERROR;
+      // If getting 500 then we will not expose about error to the user.
+      message = "Internal Server Error";
+      logger.error(e.getMessage());
+      printStackTrace(logger, e);
```

Comment:
why not log with the exception and let the logger handle it rendering the stack trace?

Reply:
`printStackTrace` is our own function where we are logging stack trace as error using LOGGER and some extra stack trace as warn. 
```
public static void printStackTrace(Logger logger, Throwable t) {
    StackTraceElement[] stack = t.getStackTrace();
    logger.error("Stacktrace with {} elements for {}", stack.length, t);
    int n = 0;
    boolean isLongStack = stack.length > STACKTRACE_LENGTH;
    if (isLongStack) {
      for (; n < STACKTRACE_LENGTH + 1; ++n) {
        logger.warn("{}: {}", n, stack[n].toString());
      }
    }
    for (; n < stack.length; ++n) {
      if (stack[n].getClassName().startsWith("ai.verta") || !isLongStack) {
        logger.warn("{}: {}", n, stack[n].toString());
      }
    }
  }
```

Reply:
if it's a common pattern to log the message, then call this, we should make a function that does that, rather than have to do this everywhere.